File: service/src/main/java/com/boxfox/cross/service/wallet/WalletService.java
package com.boxfox.cross.service.wallet;

import static com.boxfox.cross.service.wallet.indexing.IndexingMessage.EVENT_SUBJECT;
import static io.one.sys.db.tables.Wallet.WALLET;

import com.boxfox.cross.common.data.PostgresConfig;
import com.boxfox.cross.common.secure.AES256;
import com.boxfox.cross.common.vertx.service.AbstractService;
import com.boxfox.cross.service.AddressService;
import com.boxfox.cross.service.wallet.indexing.IndexingMessage;
import com.boxfox.cross.service.wallet.indexing.IndexingService;
import com.boxfox.cross.service.wallet.model.TransactionResult;
import com.boxfox.cross.service.wallet.model.TransactionStatus;
import com.boxfox.cross.service.wallet.model.WalletCreateResult;
import com.google.gson.Gson;
import io.one.sys.db.tables.daos.AccountDao;
import io.vertx.core.AsyncResult;
import io.vertx.core.Future;
import io.vertx.core.Handler;
import io.vertx.core.Vertx;

import java.util.List;

public abstract class WalletService extends AbstractService{
  private IndexingService indexingService;
  private String symbol;

  public WalletService(Vertx vertx, String symbol){
    super(vertx);
    this.symbol = symbol;
  }

  public abstract String getBalance(String address);

  public final void createWallet( String uid, String name,String password, String description, Handler<AsyncResult<String>> res){
    doAsync(future -> {
      WalletCreateResult walletCreateResult = createWallet(password);
      if(walletCreateResult.isSuccess()) {
        String walletFileName = AES256.encrypt(walletCreateResult.getWalletName());
        AccountDao dao = new AccountDao(PostgresConfig.create());
        useContext(ctx -> {
          ctx.insertInto(WALLET).values(uid, symbol.toUpperCase(), name, description, walletCreateResult.getAddress(), AddressService.createRandomAddress(dao), walletFileName).execute();
        });
        future.complete(walletCreateResult.getAddress());
      }else{
        future.fail("Wallet create fail");
      }
    }, res);
  }

  public abstract WalletCreateResult createWallet(String password);

  public abstract double getPrice(String address);

  public abstract TransactionResult send(String walletFileName, String walletJsonFile, String password, String targetAddress, String amount);

  public abstract Future<List<TransactionStatus>> getTransactionList(String address) throws WalletServiceException;

  public abstract TransactionStatus getTransaction(String transactionHash) throws WalletServiceException;

  public abstract int getTransactionCount(String address) throws WalletServiceException;

  public IndexingService getIndexingService(){ return this.indexingService; }

  public void setIndexingService(IndexingService indexingService) {
    this.indexingService = indexingService;
  }

  public void indexingTransactions(String address){
    IndexingMessage msg = new IndexingMessage();
    msg.setSymbol(symbol);
    msg.setAddress(address);
    getVertx().eventBus().publish(EVENT_SUBJECT, new Gson().toJson(msg));
  }
}

File: service/src/main/java/com/boxfox/cross/service/wallet/WalletService.kt
package com.boxfox.cross.service.wallet


import com.boxfox.cross.entity.wallet.WalletCreateModel
import com.boxfox.cross.service.JooqReactiveService
import com.linkbit.android.entity.WalletModel
import io.reactivex.Completable
import io.reactivex.Single

class WalletService(private val impl: WalletServiceImpl = WalletServiceImpl()) : JooqReactiveService() {

    fun createWallet(uid: String,
                     symbol: String,
                     password: String,
                     name: String,
                     description: String,
                     open: Boolean,
                     major: Boolean): Single<WalletCreateModel> = createSingle { impl.createWallet(it, uid, symbol, password, name, description, open, major) }

    fun getWalletList(uid: String): Single<WalletModel> = createSingle { impl.getWalletList(it, uid) }

    fun findByAddress(address: String): Single<WalletModel> = createSingle { impl.getWallet(it, address) }

    fun getMajorWallet(uid: String, symbol: String): Single<WalletModel> = createSingle { impl.getMajorWallet(it, uid, symbol) }

    fun setMajorWallet(uid: String, symbol: String, address: String): Completable = createCompletable { impl.setMajorWallet(it, uid, symbol, address) }

    fun updateWallet(uid: String, address: String, name: String, description: String, open: Boolean, major: Boolean): Completable = createCompletable { impl.updateWallet(it, uid, address, name, description, open, major) }

    fun deleteWallet(uid: String, address: String): Completable = createCompletable { impl.deleteWallet(it, uid, address) }

    fun checkOwner(uid: String, address: String): Completable = createCompletable { impl.checkOwner(it, uid, address) }

    fun getBalance(symbol:String, address: String): Single<Double> = createSingle { impl.getBalance(it, symbol, address) }

    fun getBalance(address: String): Single<Double> = createSingle { impl.getBalance(it, address) }

    fun getTotalBalance(uid: String, symbol: String): Single<Double> = createSingle { impl.getTotalBalance(it, uid, symbol) }
}
